<a href="https://colab.research.google.com/github/ihebakermi10/Chatbot/blob/main/google_palm_chatbot.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Basic working of Google Palm LLM in LangChain

In [ ]:


import csv



csv_file_path = 'questions_reponses_digital_rogue_wave.csv'

with open(csv_file_path, mode='w', newline='', encoding='utf-8') as file:
    writer = csv.writer(file, delimiter=',', quotechar='"', quoting=csv.QUOTE_MINIMAL)
    writer.writerow(['prompt', 'reponse'])  # Writing the header
    for question, reponse in data:
        writer.writerow([question, reponse])


In [ ]:
%pip install langchain==0.0.284
%pip install python-dotenv==1.0.0
%pip install streamlit==1.22.0
%pip install tiktoken==0.4.0
%pip install faiss-cpu==1.7.4
%pip install protobuf~=3.19.0

In [ ]:
from langchain.llms import GooglePalm


llm = GooglePalm(google_api_key=api_key, temperature=0.1)

In [ ]:
from langchain.chains import RetrievalQA


from langchain.embeddings import GooglePalmEmbeddings
from langchain.llms import GooglePalm

### Now let's load data from Codebasics FAQ csv file

In [ ]:
from langchain.document_loaders.csv_loader import CSVLoader

loader = CSVLoader(file_path='/content/questions_reponses_digital_rogue_wave.csv', source_column="prompt")

data = loader.load()

*italicized text*### Hugging Face Embeddings

In [ ]:
print(data)

[Document(page_content='prompt: Digital Rogue Wave Maintenance Services?\nreponse: The maintenance fees of 100 DT cover bug resolution, security updates, and continuous optimization.', metadata={'source': 'Digital Rogue Wave Maintenance Services?', 'row': 0}), Document(page_content='prompt: Quick Response to Reported Issues?\nreponse: High priority for clients under daily maintenance contracts with fast issue resolution.', metadata={'source': 'Quick Response to Reported Issues?', 'row': 1}), Document(page_content='prompt: Long-term Maintenance Packages?\nreponse: Yes, packages offering rate discounts or extended services to meet various needs.', metadata={'source': 'Long-term Maintenance Packages?', 'row': 2}), Document(page_content='prompt: Flexibility of Maintenance Contracts?\nreponse: Flexible contracts, from short-term to extended commitments, tailored to the unique needs of each client.', metadata={'source': 'Flexibility of Maintenance Contracts?', 'row': 3}), Document(page_conte

In [ ]:
%pip  install InstructorEmbedding

In [ ]:
%pip install sentence-transformers


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 86.0/86.0 kB 1.8 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 35.6 MB/s eta 0:00:00
  Created wheel for sentence-transformers: filename=sentence_transformers-2.2.2-py3-none-any.whl size=125923 sha256=be9ba767ee59120e2a99422e499f3d27cd8677adbfe4c14df27a9c8a3055c745
  Stored in directory: /root/.cache/pip/wheels/62/f2/10/1e606fd5f02395388f74e7462910fe851042f97238cbbd902f
Successfully built sentence-transformers


In [ ]:
from langchain.embeddings import HuggingFaceInstructEmbeddings

instructor_embeddings = HuggingFaceInstructEmbeddings(model_name="hkunlp/instructor-large")

e = instructor_embeddings.embed_query("What is your refund policy?")

.gitattributes:   0%|          | 0.00/1.48k [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/270 [00:00<?, ?B/s]

2_Dense/config.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/3.15M [00:00<?, ?B/s]

README.md:   0%|          | 0.00/66.3k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/1.53k [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/122 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/1.34G [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/2.20k [00:00<?, ?B/s]

spiece.model:   0%|          | 0.00/792k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/2.42M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/2.41k [00:00<?, ?B/s]

modules.json:   0%|          | 0.00/461 [00:00<?, ?B/s]

load INSTRUCTOR_Transformer
max_seq_length  512


In [ ]:
len(e)

768

In [ ]:
e[:5]

[-0.0438980758190155,
 0.007685543969273567,
 -0.009231905452907085,
 0.024496249854564667,
 0.033592287451028824]

As you can see above, embedding for a sentance "What is your refund policy" is a list of size 768. Looking at the numbers in this list, doesn't give any intuitive understanding of what it is but just assume that these numbers are capturing the meaning of "What is your refund policy". If you are curious to know about embeddings, go to youtube and search "codebasics word embeddings" and you will find bunch of videos with simple, intuitive explanations

### Vector store using FAISS

In [ ]:
from langchain.vectorstores import FAISS

# Create a FAISS instance for vector database from 'data'
vectordb = FAISS.from_documents(documents=data,
                                 embedding=instructor_embeddings)

# Create a retriever for querying the vector database
retriever = vectordb.as_retriever(score_threshold = 0.7)

In [ ]:
rdocs = retriever.get_relevant_documents("how about job placement support?")
rdocs

[Document(page_content='prompt: Long-term Maintenance Packages?\nreponse: Yes, packages offering rate discounts or extended services to meet various needs.', metadata={'source': 'Long-term Maintenance Packages?', 'row': 2}),
 Document(page_content='prompt: Hosting Services Offered by Digital Rogue Wave?\nreponse: Recommendation of reliable hosting services, close collaboration with clients to choose the best option.', metadata={'source': 'Hosting Services Offered by Digital Rogue Wave?', 'row': 27}),
 Document(page_content='prompt: E-commerce Solutions for Startup Businesses?\nreponse: Yes, offering customized solutions to establish an online presence and sell effectively.', metadata={'source': 'E-commerce Solutions for Startup Businesses?', 'row': 24}),
 Document(page_content='prompt: Approach to Continuous Maintenance After Website Launch?\nreponse: Offering maintenance contracts to ensure smooth operation, with regular updates and quick assistance.', metadata={'source': 'Approach to

As you can see above, the retriever that was created using FAISS and hugging face embedding is now capable of pulling relavant documents from our original CSV file knowledge store. This is very powerful and it will help us further in our project

##### Embeddings can be created using GooglePalm too. Also for vector database you can use chromadb as well as shown below. During our experimentation, we found hugging face embeddings and FAISS to be more appropriate for our use case

In [ ]:
#google_palm_embeddings = GooglePalmEmbeddings(google_api_key=api_key)

#from langchain.vectorstores import Chroma
#vectordb = Chroma.from_documents(data,
#                            embedding=google_palm_embeddings,
#                           persist_directory='./chromadb')


#vectordb.persist()

### Create RetrievalQA chain along with prompt template 🚀

In [ ]:
from langchain.prompts import PromptTemplate



prompt_template = """

Given the following context and a question, generate an answer based on this context only.
In the answer try to provide as much text as possible from "reponse" section in the source document context without making much changes.
If the answer is not found in the context, kindly state "I don't know." Don't try to make up an answer.

CONTEXT: {context}
QUESTION: {question}

"""


PROMPT = PromptTemplate(
    template=prompt_template, input_variables=["context", "question"]
)
chain_type_kwargs = {"prompt": PROMPT}


from langchain.chains import RetrievalQA

chain = RetrievalQA.from_chain_type(llm=llm,
                            chain_type="stuff",
                            retriever=retriever,
                            input_key="query",
                            return_source_documents=True,
                            chain_type_kwargs=chain_type_kwargs)


### We are all set 👍🏼 Let's ask some questions now

**As you can see above, the answer of question comes from two different FAQs within our csv file and it is able to pull those questions and merge them nicely**

In [ ]:
chain("les etape dengagement et la dure?")

{'query': 'les etape dengagement et la dure?',
 'result': 'Flexible contracts, from short-term to extended commitments, tailored to the unique needs of each client.',
 'source_documents': [Document(page_content='prompt: Collaboration Process with Clients at Digital Rogue Wave?\nreponse: In-depth consultation to understand goals and expectations, open communication throughout the project.', metadata={'source': 'Collaboration Process with Clients at Digital Rogue Wave?', 'row': 21}),
  Document(page_content='prompt: Initiating a Daily Maintenance Contract?\nreponse: Contact our team with your site details to start the registration process.', metadata={'source': 'Initiating a Daily Maintenance Contract?', 'row': 6}),
  Document(page_content='prompt: Long-term Maintenance Packages?\nreponse: Yes, packages offering rate discounts or extended services to meet various needs.', metadata={'source': 'Long-term Maintenance Packages?', 'row': 2}),
  Document(page_content='prompt: Flexibility of Ma

In [ ]:
from langchain.vectorstores import FAISS
from langchain.llms import GooglePalm
from langchain.document_loaders.csv_loader import CSVLoader
from langchain.embeddings import HuggingFaceInstructEmbeddings
from langchain.prompts import PromptTemplate
from langchain.chains import RetrievalQA
import os

from dotenv import load_dotenv
load_dotenv()  # take environment variables from .env (especially openai api key)

# Create Google Palm LLM model
llm = GooglePalm(google_api_key="*****", temperature=0.1)
# # Initialize instructor embeddings using the Hugging Face model
instructor_embeddings = HuggingFaceInstructEmbeddings(model_name="hkunlp/instructor-large")
vectordb_file_path = "faiss_index"

def create_vector_db():
    # Load data from FAQ sheet
    loader = CSVLoader(file_path='/content/questions_reponses_digital_rogue_wave.csv', source_column="prompt")
    data = loader.load()

    # Create a FAISS instance for vector database from 'data'
    vectordb = FAISS.from_documents(documents=data,
                                    embedding=instructor_embeddings)

    # Save vector database locally
    vectordb.save_local(vectordb_file_path)


def get_qa_chain():
    # Load the vector database from the local folder
    vectordb = FAISS.load_local(vectordb_file_path, instructor_embeddings)

    # Create a retriever for querying the vector database
    retriever = vectordb.as_retriever(score_threshold=0.7)

    prompt_template = """Given the following context and a question, generate an answer based on this context only.
    In the answer try to provide as much text as possible from "response" section in the source document context without making much changes.
    If the answer is not found in the context, kindly state "I don't know." Don't try to make up an answer.

    CONTEXT: {context}

    QUESTION: {question}"""

    PROMPT = PromptTemplate(
        template=prompt_template, input_variables=["context", "question"]
    )

    chain = RetrievalQA.from_chain_type(llm=llm,
                                        chain_type="stuff",
                                        retriever=retriever,
                                        input_key="query",
                                        return_source_documents=True,
                                        chain_type_kwargs={"prompt": PROMPT})

    return chain

if __name__ == "__main__":
    create_vector_db()
    chain = get_qa_chain()
    print(chain("cost ?"))

load INSTRUCTOR_Transformer
max_seq_length  512
{'query': 'cost ?', 'result': 'The maintenance fees of 100 DT cover bug resolution, security updates, and continuous optimization.', 'source_documents': [Document(page_content='prompt: Digital Rogue Wave Maintenance Services?\nreponse: The maintenance fees of 100 DT cover bug resolution, security updates, and continuous optimization.', metadata={'source': 'Digital Rogue Wave Maintenance Services?', 'row': 0}), Document(page_content='prompt: Long-term Maintenance Packages?\nreponse: Yes, packages offering rate discounts or extended services to meet various needs.', metadata={'source': 'Long-term Maintenance Packages?', 'row': 2}), Document(page_content='prompt: Periodic Adjustments to Maintenance Fees?\nreponse: Stable rates, periodic evaluation to stay competitive, transparent adjustments.', metadata={'source': 'Periodic Adjustments to Maintenance Fees?', 'row': 7}), Document(page_content='prompt: Initiating a Daily Maintenance Contract?\

In [ ]:
%pip install streamlit


In [ ]:
! pip install streamlit -q
!wget -q -O - ipv4.icanhazip.com
! streamlit run /content/sample_data/main.py & npx localtunnel --port 8501

35.199.16.129
npx: installed 22 in 3.289s



  You can now view your Streamlit app in your browser.

  Network URL: http://172.28.0.12:8501
  External URL: http://35.199.16.129:8501

your url is: https://sour-women-smile.loca.lt
load INSTRUCTOR_Transformer
max_seq_length  512
